runtime/src/iree/builtins/ukernel/tools/pack_test.c

```diff
@@ -214,8 +214,12 @@ int main(int argc, char** argv) {
 #elif defined(IREE_UK_ARCH_X86_64)
   iree_uk_test_pack(iree_uk_pack_type_f32f32, 8, 1, cpu->avx2_fma);
   iree_uk_test_pack(iree_uk_pack_type_i8i8, 8, 2, cpu->avx2_fma);
+  iree_uk_test_pack(iree_uk_pack_type_f32f32, 8, 8, cpu->avx2_fma);
+  iree_uk_test_pack(iree_uk_pack_type_i32i32, 8, 8, cpu->avx2_fma);
   iree_uk_test_pack(iree_uk_pack_type_f32f32, 16, 1, cpu->avx512_base);
   iree_uk_test_pack(iree_uk_pack_type_i8i8, 16, 2, cpu->avx512_base);
+  iree_uk_test_pack(iree_uk_pack_type_f32f32, 16, 16, cpu->avx512_base);
```

Comment:
What are all the valid configs for the different pack types?

Reply:
This parameters struct, https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/pack.h#L30-L46 fully describes the interface since the ukernel entry point [`iree_uk_pack`](https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/pack.h#L63-L64) takes nothing but a pointer to that.

The range of valid values for these struct fields is whatever is accepted by the validation function (`iree_uk_pack_validate`), which runs a bunch of assertions (`IREE_UK_ASSERT`) on these values. (So validation evaporates in non-assertion release builds; it is the caller's responsibility to make correct calls, validation is done only to catch programming mistakes, not as part of a contract):
https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/pack.c#L55-L100

So in particular, the valid pack types are the 3 enum values accepted here
https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/pack.c#L60-L62

And that's really all 3 enumerated values for this enum,
https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/pack.h#L16-L20

If you want to find a reference implementation of a ukernel to understand the semantics of all these parameters, head over to the unit test, which has got to contain its own reference implementation to test against, here `iree_uk_pack_reference`:
https://github.com/openxla/iree/blob/6dcca4094988af88035fe72159df62753c790b59/runtime/src/iree/builtins/ukernel/tools/pack_test.c#L12-L56

Reply:
Thanks!